### PR TITLE
Speed ups for navigating Microsoft Excel spreadsheets, especially for those containing comments / dropdowns 

### LINKED ISSUE #7348: Extremely slow navigating between cells in MS Excel 2016 when a cell contains a comment or dropdown
when using MS office 2016 as a direct enterprise install, or MS office 365, I notice a real slow down when using ms excel and using the arrow keys to navigate between cells. Sometimes it can take up to 2 seconds to move between cells. I have seen this issue on many different types of coputer. I have attached an excell file which you can see the problem in. with office 2010, this does not appear to be an issue. sometimes there is no speech at all when moving between cells, particularly if i have entered something into a cell and then hit enter to update it. I have noticed this is particularly present when inputting numbers, which you can do by using the price column in my excel sheet i've attached. The only way to fix this seems to alt tab out of the excel window and then back in again before it starts reading the cells, all be it very slowly. Incidentely I just have it reading cell coordinates, not boarders or other information beyond the text in the cells.
I have noticed this issue for about the last 8 months using the NVDA next builds, but just haven't gotten around to reporting it. My bad.
I use Excel daily for work now so this is getting to be a real issue for me. Any help you could offer would be most appreciated and I am happy to try things to help. many Thanks.
[collection.xlsx](https://github.com/nvaccess/nvda/files/1118055/collection.xlsx)

### LINKED ISSUE #8271: Arrowing around this worksheet has 2 seconds of lag
### Steps to reproduce:
1. [Download this workbook](https://www.dropbox.com/s/jd3i2ja904dgyta/excel%20lag.xlsx?dl=0)
2. Open it in excel and arrow around.

### Expected behavior:
There should be no lag when moving around the cells.

### Actual behavior:
About 2 seconds of lag every time you move to a new cell.

### System configuration:
NVDA version: master-15118,2229d55a

NVDA Installed or portable: installed

Other information:

Windows version: windows 10 1803

Name and version of other software in use when reproducing the issue: Excel® 2016 MSO (16.0.9226.2156

### Other questions:

Does the issue still occur after restarting your PC? Yes

Have you tried any other versions of NVDA? No

Other notes: Narrator does not lag at all, maybe because its using UIA?

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #8271 
Fixes #7348 

### Summary of the issue:
As NVDA's support for Microsoft Excel has become more rich over the years, the amount of cross-process COM calls NVDA makes in order to report a cell has significantly increased. these days, at least 65 cross-process COM calls are made to report one cell. A large chunk of these calls are to do with calculating overflowing/chropping info. In general the increasing number of calls is causing responsiveness in Excel to slow down. However, due to a serious bug in Microsoft Excel 2016 (and possibly in 2013) if a spreadsheet contains at least one comment or at least one validation dropdown list, all cross-process COM calls increase in time, usually causing responsiveness to slow down by around a factor of 7. for some spreadsheets this can mean the NVDA user is waiting up to 1.5 seconds or more when navigating from one cell to another.
NV Access has been working with Microsoft to try and solve the underlying bug in Excel itself, and there may be a fix coming from Microsoft's side in the near future. However, existing builds of Microsoft Excel would not likely get this fix.
 
### Description of how this pull request fixes the issue:
Some helper functions have been added to nvdahelper inproc utils which allows NVDA to fetch all required information for an Excel cell in one cross-process call. This totally removes the lag seen in spreadsheets containing comments and d or dropdowns, and may provide an improvement to other spreadsheets, especially if we start adding more features again in the future.
A small speed up has also been achieved when listing comments and formulas in the NVDA Elements list, again by fetching all information in one call.

### Testing performed:
Tbd. Requires plenty :)
 
### Known issues with pull request:
As NVDA cannot inject into Excel's process when a spreadsheet is opened in Protected View, we can only support very basic information such as a cell's address and content. Fetching further information would simply cause the original bug. At least this way a basic information can be read in a Spreadsheet in Protected view. Previously performance was so bad it was pretty much pointless to try anyway.
 
### Change log entry:
Bug fixes:
- Significant performance improvements when navigating cells in Microsoft Excel, particularly when the spreadsheet contains comments and or validation dropdown lists. (#7348)

### CODE REVIEW

File: source/NVDAObjects/window/excel.py
Comment (leonardder) [inline source/NVDAObjects/window/excel.py]: Could you add abc.ABCMeta as a metaclass and then decorate this with abc.abstractmethod?

File: source/NVDAObjects/window/excel.py
Comment (leonardder) [inline source/NVDAObjects/window/excel.py]: I assume this has to be QuickNavItemClass instead?

**Review by LeonarddeR — CHANGES_REQUESTED** [review]

File: nvdaHelper/remote/excel.cpp
Comment (leonardder) [inline nvdaHelper/remote/excel.cpp]: Github somehow decided to discard my comment here.
Since states is supposed to be a bitmask, I'd personally prefer `states|=` over `states+=` for this and all subsequent cases.

File: nvdaHelper/remote/excel.cpp
Comment (feerrenrut) [inline nvdaHelper/remote/excel.cpp]: Many (most, all?) are not checking the return result. Please add:
```suggestion
	auto res = _com_dispatch_raw_propget(pDispatchRange,XLDISPID_RANGE_TEXT,VT_BSTR,&sText);
	if( res !=S_OK){
		LOG_DEBUGWARNING(L"range text failed");
		return 0;
	}
```

File: nvdaHelper/remote/excel.cpp
Comment (feerrenrut) [reply nvdaHelper/remote/excel.cpp]: Yes, I think that using `+=` leads to a logic error.

File: nvdaHelper/remote/excel.cpp
Comment (feerrenrut) [inline nvdaHelper/remote/excel.cpp]: why is this commented out?

**Review by feerrenrut — CHANGES_REQUESTED** [review]
I've given this a once over. There are a few little things to fix up. It's probably worth getting this into the hands of someone who is using excel a lot to help us test it.

File: nvdaHelper/remote/excel.cpp
Comment (michaelDCurran) [reply nvdaHelper/remote/excel.cpp]: Done.

File: nvdaHelper/common/COMUtils.h
Comment (michaelDCurran) [reply nvdaHelper/common/COMUtils.h]: Yep. Done and tested.

File: source/NVDAObjects/window/excel.py
Comment (michaelDCurran) [reply source/NVDAObjects/window/excel.py]: I'm still unsure as to how best to represent selection in Excel from a user experience point of view. Until we can settle on something good, I don't see any reason to change the current code.

File: source/NVDAObjects/window/excel.py
Comment (michaelDCurran) [reply source/NVDAObjects/window/excel.py]: Done.

File: source/NVDAObjects/window/excel.py
Comment (michaelDCurran) [reply source/NVDAObjects/window/excel.py]: Yes. But it has now been replaced by an abstract property anyway.

File: source/NVDAObjects/window/excel.py
Comment (michaelDCurran) [reply source/NVDAObjects/window/excel.py]: I have provided detailed comments for the reason and the address arguments.

File: source/NVDAObjects/window/excel.py
Comment (michaelDCurran) [reply source/NVDAObjects/window/excel.py]: No... I have now put it back :)

File: nvdaHelper/remote/excel.cpp
Comment (michaelDCurran) [reply nvdaHelper/remote/excel.cpp]: Done. and Also addressed better error catching here for @feerrenrut

File: nvdaHelper/remote/excel.cpp
Comment (michaelDCurran) [reply nvdaHelper/remote/excel.cpp]: Done.

File: nvdaHelper/remote/excel.cpp
Comment (michaelDCurran) [reply nvdaHelper/remote/excel.cpp]: It always produces an error. I have now suppressed this log call only for that particular error, and commented why.

**Comment by michaelDCurran** [comment]
@leonardder  and @feerrenrut  I believe I have addressed all your review comments. A try build has been linked to on the two related issues.

File: source/NVDAObjects/window/excel.py
Comment (leonardder) [reply source/NVDAObjects/window/excel.py]: See also #6959 and #5401

File: nvdaHelper/remote/excel.cpp
Comment (leonardder) [inline nvdaHelper/remote/excel.cpp]: The results of this com call and the calls below are not yet logged on failure.

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Looks good to me now, just one comment. Also, I agree with @feerrenrut that this needs some testing by more frequent Excel users. However, I think Alpha might be a good stage to do this

**Comment by dnz3d4c** [comment]
I hope this pr is merged into the alpha version. Because it can be tested by more users. Testing with try build is inconvenient.

> 
> 
> Looks good to me now, just one comment. Also, I agree with @feerrenrut that this needs some testing by more frequent Excel users. However, I think Alpha might be a good stage to do this

**Comment by codeofdusk** [comment]
Does this PR close the first idea on [this page](https://github.com/nvaccess/nvda/wiki/GSOC-2019-Ideas-List)? If so that page should probably be edited accordingly.

**Comment by michaelDCurran** [comment]
@codeofdusk This pr does not go as far as it can. E.g. pressing NVDA+f to fetch formatting, or if any of the formatting checkboxes are on in NVDA's document settings, Excel will still be very slow. This pr just speeds up the most common usage of navigating around cells with default settings. However, once the pr is merged I will be sure to edit that page. Particular factors meant that NV Access needed to give this particular bug much higher priority right now.

File: nvdaHelper/remote/excel.cpp
Comment (michaelDCurran) [reply nvdaHelper/remote/excel.cpp]: All COM calls now check for errors. Plus I removed a duplicate call to range.address which I'd accidentally left in after a previous refactor.

**Comment by michaelDCurran** [comment]
Although I'd like to merge this to alpha to get wide testing. @feerrenrut  and I have discussed about using this codebase to fuly test using the auto-generated MS Office C++ COM interfaces. This code base is much smaller than the MS Word code so shouldn't take me too long to convert. However, as there are still some open issues around using the COM interfaces rather than the raw IDispatch helper functions, I wouldn't want to hold this current pr up for that.

**Comment by michaelDCurran** [comment]
Here is the latest try build: https://ci.appveyor.com/api/buildjobs/u3cvt0l42g757srt/artifacts/output%2Fnvda_snapshot_try-excelInproc-16752%2C35b618a6.exe
I'm extremely interested to know if there are any issues with this on early versions of Office like 2010 or 2007.

**Comment by leonardder** [comment]
> Although I'd like to merge this to alpha to get wide testing. @feerrenrut and I have discussed about using this codebase to fully test using the auto-generated MS Office C++ COM interfaces. This code base is much smaller than the MS Word code so shouldn't take me too long to convert. However, as there are still some open issues around using the COM interfaces rather than the raw IDispatch helper functions, I wouldn't want to hold this current pr up for that.

I tend to agree. Are the auto-generated MS Office C++ COM interfaces supposed to be faster? I guess we have to double check that before changing the Word code in such a major way.

**Review by LeonarddeR — APPROVED** [review]

**Comment by michaelDCurran** [comment]
Actually, after playing with the auto-generated C++ COM interfaces for a while today I'm not convinced this is a good idea. At least not for Excel anyway. The main reason is that many of the methods are declared taking VARIANTs as arguments. Even things like range.text, font.size etc are all variants. This means that you need to create and initialize a variant for each argument, execute the call, and then possibly check and fetch the real value from any [out] variants. This therefore increases the complexity and amount of lines for each call significantly above what we currently have with the _com_dispatch_raw_methodcalls, as they automatically handle converting to/from the real native types. Note for MS Word, they havn't taken this kind of shortcut and therefore in future there may be still some advantage to consider doing it for MS word.
However, perhaps we can at least write some neicer wrapper functions for _com_dispatch_raw_* functions,  somehow we could instead of declaring the dispids by themselves, instead declare the dispid along with its expected argument types, giving it a friendly name, and then using that in the actual code. 
But even doing that, I'm not sure that would be of high priority.

**Comment by lukaszgo1** [comment]
I've tested on Excel 2010 and one problem which I've noticed is the fact that on NVDA's element list some cells containing text are recognized as containing formulas. It mainly happens when the cell is a header and cells below are containing actual formulas. I am attaching a very simple sheet in which it can be reproduced. The cell c1 is recognized as containing formula whereas it contain only word test.
[Excel-tests.xlsx](https://github.com/nvaccess/nvda/files/2862423/Excel-tests.xlsx)

**Comment by michaelDCurran** [comment]
Thanks for reporting the incorrect formula listing. In fact you probably 
would have found that the last formula was not listed either. I.e. 
everything was off by 1.

I have fixed and verified this now.

A try build is coming soon.

**Comment by michaelDCurran** [comment]
Here is the latest try build: https://ci.appveyor.com/api/buildjobs/746a4sf5p0di7rcd/artifacts/output%2Fnvda_snapshot_try-excelInproc-16767%2C5195b108.exe
This fixes the issue where listing formulas in the Elements list would be including the header cell above, and missing the last formula in the sheet.

**Review by LeonarddeR — APPROVED** [review]

**Comment by lukaszgo1** [comment]
While the formula at the top of the column is no longer shown in the elements  list there are still some empty cells which are displayed as having formula in the element list. Unfortunately I cannot share the workbook in which I have this issue here. Is there anyway to send it to you privately?

**Review by feerrenrut — APPROVED** [review]

**Comment by michaelDCurran** [comment]
Please email the spreadsheet you are having the issue with to 
info@nvaccess.org. Thanks.

**Comment by michaelDCurran** [comment]
Here is another try build: 
https://ci.appveyor.com/api/buildjobs/7dco6i3ccni4ehje/artifacts/output%2Fnvda_snapshot_try-excelInproc-16772%2Cc539f98b.exe

My testing with the spreadsheet sent privately now shows the unwanted 
blank cells are no longer being listed as formulas in the Elements list.

Technical:

Iterating over an Excel range using range.item is not what we want as if 
the range is filtered (I.e. is made up of noncontinuous subranges), item 
may return cells that fall within the gaps. Rather we must use the 
IEnumVARIANT interface on the range's _newenum property. This is what 
comtypes did internally when iterating over a range in NVDA out-of-process.

**Comment by lukaszgo1** [comment]
I can confirm it's fixed for me in Excel 2010 with the same sheet.

**Pull request merged**
